[GlobalISel] Check whether `G_CTLZ` is legal in `matchUMulHToLShr`

shiltian opened the pull request:
We need to check `G_CTLZ` because the combine uses `G_CTLZ` to get log base 2,
and it is not always legal for on a target.

Fixes SWDEV-512440.

shiltian commented:
* **#126457** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/126457?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/126457?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

arsenm reviewed (approved):
Can you also add and end to end IR test where this reproduced

shiltian committed:
[GlobalISel] Check whether `G_CTLZ` is legal in `matchUMulHToLShr`

We need to check `G_CTLZ` because the combine uses `G_CTLZ` to get log base 2,
and it is not always legal for on a target.

shiltian commented on llvm/test/CodeGen/AMDGPU/GlobalISel/no-ctlz-from-umul-to-lshr-in-postlegalizer.ll:
```diff
@@ -0,0 +1,98 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
There are two interesting things in this test case:

1. It is not reproducible w/o `-O0` argument
2. The `udiv` is actually a UB, and replacing either of its operand will make the crash go away (or the pattern not recognized such that the issue will not happen).

arsenm replied:
You would have an easier time reproducing this by directly emitting a multiply + shift by half bitwidth. it just happens one appears inside the div expansion

shiltian replied:
like this? it doesn't seem to be able to reproduce the issue.

```
define void @test(ptr %p, i32 %a, i32 %b) {
  %mul = mul i32 %a, %b
  %lshr = lshr i32 %mul, 16
  store i32 %lshr, ptr %p, align 4
  ret void
}
```

arsenm replied:
You probably want 64-bit multiply, but I don't see the general combine to form umulh from mul + shift
